src/thumbnailer_slope_optim.cc

```diff
@@ -0,0 +1,253 @@
+  }
+
+  // Use binary search with slope optimization to find quality values that makes
+  // the animation fit the given byte budget. The quality value for each frame
+  // can be different.
+  while (min_quality <= max_quality && !optim_list.empty()) {
+    int mid_quality = (min_quality + max_quality) / 2;
+
+    std::queue<int> new_optim_list;
+
+    while (!optim_list.empty()) {
+      int curr_frame = optim_list.front();
+      optim_list.pop();
+
+      float curr_slope;
+      if (ComputeSlope(curr_frame, min_quality, max_quality, &curr_slope) !=
+          kOk) {
+        return kSlopeOptimError;
+      }
+
+      if (frames_[curr_frame].final_quality == -1 || curr_slope > limit_slope) {
+        frames_[curr_frame].config.quality = mid_quality;
+        new_optim_list.push(curr_frame);
+      }
+    }
+
+    if (new_optim_list.empty()) break;
+
+    if (GenerateAnimationNoBudget(&new_webp_data) != kOk) {
+      return kMemoryError;
+    }
+
+    if (new_webp_data.size <= byte_budget_) {
+      while (!new_optim_list.empty()) {
+        int curr_frame = new_optim_list.front();
+        new_optim_list.pop();
+        frames_[curr_frame].final_quality = mid_quality;
+        optim_list.push(curr_frame);
+      }
+      WebPDataClear(webp_data);
+      *webp_data = new_webp_data;
+      min_quality = mid_quality + 1;
+    } else {
+      optim_list = new_optim_list;
+      max_quality = mid_quality - 1;
+      WebPDataClear(&new_webp_data);
+    }
+  }
+
+  std::cout << "Final qualities with slope optimization: ";
+
+  for (auto& frame : frames_) {
+    frame.config.quality = frame.final_quality;
+    if (GetPictureStats(frame.pic, frame.config, &frame.encoded_size,
+                        &frame.final_psnr) != kOk) {
+      return kStatsError;
+    }
+    std::cerr << frame.config.quality << " ";
+  }
+
+  std::cout << std::endl;
+
+  return (webp_data->size > 0) ? kOk : kByteBudgetError;
+}
+
+Thumbnailer::Status Thumbnailer::LossyEncodeNoSlopeOptim(
+    WebPData* const webp_data) {
+  int anim_size = 0;
+  for (auto& frame : frames_) {
+    anim_size += frame.encoded_size;
+  }
+
+  WebPData new_webp_data;
+  WebPDataInit(&new_webp_data);
+
+  // For each frame, find the best quality value that can produce the higher
+  // PSNR than the current one if possible.
+  for (auto& frame : frames_) {
+    int min_quality = 0;
+    int max_quality = 100;
+    frame.config.lossless = 0;
+    while (min_quality <= max_quality) {
+      int mid_quality = (min_quality + max_quality) / 2;
+      int new_size;
+      float new_psnr;
+      frame.config.quality = mid_quality;
+      if (GetPictureStats(frame.pic, frame.config, &new_size, &new_psnr) !=
+          kOk) {
+        return kStatsError;
+      }
+
+      if (new_psnr > frame.final_psnr || ((new_psnr == frame.final_psnr) &&
+                                          (new_size <= frame.encoded_size))) {
+        if (anim_size - frame.encoded_size + new_size <= byte_budget_) {
```

Comment:
This might tend to use all the available budget for the first frames and leave no budget for the other frames. Maybe we should try to allocate some budget to every frame, like:
int extra_budget = (byte_budget_ - anim_size) / num_remaining_frames;
if (new_size - frame.encoded_size  < extra_budget)